pytorch_lightning/accelerators/accelerator.py

```diff
@@ -0,0 +1,365 @@
+from pytorch_lightning.utilities.enums import LightningEnum
+from typing import Any, Callable, Iterable, Optional, Union
+
+import torch
+from torch.optim import Optimizer
+
+from pytorch_lightning.plugins import TrainingTypePlugin, HorovodPlugin
+from pytorch_lightning.plugins.precision import (
+    ApexMixedPrecisionPlugin,
+    MixedPrecisionPlugin,
+    NativeMixedPrecisionPlugin,
+    PrecisionPlugin,
+)
+from pytorch_lightning.core import LightningModule
+from pytorch_lightning.utilities import AMPType
+from pytorch_lightning.utilities.apply_func import move_data_to_device
+
+
+class Accelerator(object):
+    """
+    The Accelerator Base Class.
+    An Accelerator is meant to deal with one type of Hardware.
+
+    Currently there are accelerators for:
+    - CPU
+    - GPU
+    - TPU
+
+    Each Accelerator gets two plugins upon initialization:
+    One to handle differences from the training routine and one to handle different precisions.
+
+    """
+
+    def __init__(
+        self,
+        precision_plugin: PrecisionPlugin,
+        training_type_plugin: TrainingTypePlugin,
+    ) -> None:
+        """
+
+        Args:
+            precision_plugin: the plugin to handle precision-specific parts
+            training_type_plugin: the plugin to handle different training routines
+        """
+        self.precision_plugin = precision_plugin
+        self.training_type_plugin = training_type_plugin
+
+        self.optimizers = None
+        self.lr_schedulers = None
+        self.optimizer_frequencies = None
+
+    def setup(self, trainer: "Trainer", model: LightningModule) -> None:
+        """
+        Connects the plugins to the training process, creates optimizers
+
+        Args:
+            trainer: the trainer instance to connect to
+            model: the model to train
+        """
+        self.connect_training_type_plugin(self.training_type_plugin, model)
+        self.setup_optimizers(trainer, model)
+        self.connect_precision_plugin(self.precision_plugin)
+        self.optimizers = trainer.convert_to_lightning_optimizers(self.optimizers)
+
+    @property
+    def model(self) -> torch.nn.Module:
+        """Returns the model. This can also be a wrapped LightningModule.
+        For retrieving the pure LightningModule use :attr:`Accelerator.lightning_module`
+
+        """
+        return self.training_type_plugin.model
+
+    @model.setter
+    def model(self, new_model: torch.nn.Module) -> None:
+        self.training_type_plugin.model = new_model
+
+    @property
+    def lightning_module(self) -> LightningModule:
+        """Returns the pure LightningModule.
+        To get the potentially wrapped model use :attr:`Accelerator.model`
+
+        """
+        return self.training_type_plugin.lightning_module
```

Comment:
when should one call `model` vs `lightning_module` ? is `model` meant to capture the wrapped module for data/distributed data parallel?

Reply:
Yes it is. So when you want the model (e.g. to forward through ddp) you call .model and when you want to access anything from the lightning module you call .lightning_module which is implemented by each trainingtype plugin and takes care of unwrapping